src/rewriter.coffee

```diff
@@ -548,20 +548,86 @@ exports.Rewriter = class Rewriter
   # `OUTDENT` tokens should always be positioned at the last character of the
   # previous token, so that AST nodes ending in an `OUTDENT` token end up with a
   # location corresponding to the last “real” token under the node.
-  fixOutdentLocationData: ->
+  fixIndentationLocationData: ->
+    @allComments ?= extractAllCommentTokens @tokens
```

Comment:
We have to extract the comment tokens here inside the rewriter as well

Since we're concerned with performance of these additional passes over the tokens (to extract comments), this starts leading me back toward the way I implemented this on my `prettier` branch:

Instead of having to re-extract comments from the tokens, just keep track of the comments separately (in addition to the current behavior of attaching comments to tokens) in the lexer and then explicitly pass that list of comments around (between the lexer and rewriter and between the lexer/rewriter and nodes tree via the return value of the lexer)

Doing it that way we wouldn't have to do any additional passes over the tokens in order to extract comments since the comment list would get build up as we go in the lexer

I'd be inclined to do a separate refactor/optimization PR where we could try this pattern out (or see what else we come up with) if that works for you

Reply:
Definitely in favor of optimization as a phase 2 task, as I’m eager to get this out the door. I also wonder just how inefficient this really is. There is the `cake bench` command; how does this branch measure against the `ast` branch (i.e. how much slower does this PR make things)? The difference between those two values is the outer limit as to how much optimization could improve things.

Reply:
Here are the results from `cake bench`:
`ast` branch:
```
Lex      630 ms (58227 tokens)
Rewrite  683 ms (64842 tokens)
Parse    977 ms
Compile  839 ms (494960 chars)
total   3129 ms
```
This branch:
```
Lex      639 ms (58914 tokens)
Rewrite  847 ms (65549 tokens)
Parse    980 ms
Compile  881 ms (501478 chars)
total   3347 ms
```
So a somewhat slower rewriter run

I believe that longer source files with lots of comments would gain the most from optimizing this rewriter pass

I'd be fine with trying to get in an optimization PR before we land `ast`, or deferring until phase 2

Reply:
Up to you. 218 ms difference is I think 6% slower, which seems pretty trivial. I think it could wait until phase 2.